### PR TITLE
Improve livecheck docs.

### PR DESCRIPTION
**Pull request by reitermarkus**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [ ] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?
- [ ] Have you successfully run `brew man` locally and committed any changes?

-----

Reworded some things, unindented examples and added docs for `Sparkle` `strategy` blocks.

### CODE REVIEW

**Comment by reitermarkus** [comment]
I also removed the `regex` from the block examples. I think it should be **either** a `regex` **or** a block, not both, especially since some blocks require more than one regex.

cc @samford

**Comment by samford** [comment]
> I also removed the `regex` from the block examples. I think it should be **either** a `regex` **or** a block, not both, especially since some blocks require more than one regex. I didn't even know this functionality was introduced before reading through this.

I specifically added the ability to pass in the regex to the strategy block (whether it's defined using `#regex` or simply passing in the generated regex from the strategy) because it allows us to continue surfacing the regex in the debug and JSON output.

In instances, where we only use one regex in the `strategy` block, we should define the regex before the `strategy` block and simply pass it in, as shown in the documentation examples. In instances where more than one regex is used, we can simply define/use the regexes in the strategy block. There are some instances where we use two regexes and it's appropriate to define the main one using `#regex` and define a supplementary one in the `strategy` block.

I think always defining regexes within the `strategy` block is less useful, as we can't get at that information in the debug and JSON output. Using `#regex` is more informative/useful and in keeping with what we're doing outside of `strategy` blocks.

File: docs/Brew-Livecheck.md
Comment (samford) [inline docs/Brew-Livecheck.md:14]:
```suggestion
It's sometimes necessary to override this default behavior to create a working check. If a source doesn't provide the newest version, we need to check a different one. If livecheck doesn't correctly match version text, we need to provide an appropriate regex or `strategy` block.
```

"block" can be ambiguous here for someone who's unfamiliar with livecheck, so it wouldn't hurt to make it clear.

File: docs/Brew-Livecheck.md
Comment (samford) [inline docs/Brew-Livecheck.md:68]: The reason for the indentation here is so that you can easily copy and paste this text into a formula/cask. Depending on your text editor, if there's no leading spaces here you have to manually indent the lines whenever you copy and paste it. There's no harm in having the spaces but removing them would make it less beneficial, so I think it should be kept like this to make it easy for contributors/maintainers.

File: docs/Brew-Livecheck.md
Comment (samford) [inline docs/Brew-Livecheck.md:101]: As mentioned in [my comment on the docs PR](https://github.com/Homebrew/brew/pull/10329#issuecomment-760332137), I don't think it should be an either/or situation and I specifically added the ability for us to pass the `regex` into `strategy` blocks for a good reason.

In instances where we're only using one regex in the `strategy` block, it allows us to continue to surface the regex in the debug and JSON output, which is useful. When the regex isn't specified, it will also pass in a regex that's generated within the strategy (i.e., the strategy regex), so we can simply use that in the block instead of having to create a regex literal to replicate it.

There are benefits to specifying the regex outside of the strategy block but I don't see any benefit to always specifying the regex(es) within the strategy block. It unnecessarily prevents us from surfacing that information in a useful manner in the output.

File: docs/Brew-Livecheck.md
Comment (samford) [inline docs/Brew-Livecheck.md:121]: I have the commentary after the code block in the other examples, so we should standardize on having it either before or after for all the examples.

**Comment by reitermarkus** [comment]
For what it's worth, I have not once looked at the regex in the debug output for debugging the regex.

File: docs/Brew-Livecheck.md
Comment (samford) [inline docs/Brew-Livecheck.md:140]:
```suggestion
  url "https://www.example.com/example.xml"
```

I tend to use example URLs in the documentation, to avoid any potential issues that may arise from using a live URL (not from example.com or a domain under our control).

**Review by samford — DISMISSED** [review]

**Comment by samford** [comment]
> For what it's worth, I have not once looked at the regex in the debug output for debugging the regex. Also not sure what use it serves in the JSON output.

That's fine but other people, like myself, do reference it in the debug output on a regular basis.

Can you explain the benefit(s) of always defining the regex(es) in `strategy` blocks? From my perspective, I can't see any benefit, whereas there are benefits to defining the regex outside of the `strategy` block when possible.

**Comment by reitermarkus** [comment]
> Can you explain the benefit(s) of always defining the regex(es) in strategy blocks? From my perspective, I can't see any benefit, whereas there are benefits to defining the regex outside of the strategy block when possible.

It's more readable because all logic is in a single block of code. It's not clear that the regex that is passed into the block is the same as the one outside the block by just looking at it. Also, you have to write `regex` three times instead of simply inlining the regex straight away. It's also generally one line longer. Lastly, it will lead to inconsistent usage anyways given there are blocks that need more than one regex.

**Comment by samford** [comment]
> It's more readable because all logic is in a single block of code.

The regex is directly above the `strategy` block, so I don't think using `#regex` would dramatically affect this in a negative way. Some of the regexes I've seen in casks are very long, so the shorter `regex` variable name could potentially make the code easier to read in those instances.

> It's not clear that the regex that is passed into the block is the same as the one outside the block by just looking at it.

If you feel there's ambiguity about this, then we should explain the behavior in the documentation to make it clear. I don't foresee this being a common issue in practice.

Everyone who's working with `livecheck` blocks should already understand the fundamental concept that the regex supplied in the `livecheck` block is provided to the strategy. `strategy` blocks are simply a replacement of the strategy's matching logic, so it follows that the regex passed to the `strategy` block is the regex provided in the `livecheck` block. Even if someone doesn't understand this, the `#regex` call appears immediately before the `strategy` block, so I don't think this will confuse anyone.

We've already had a contributor in homebrew/core, who only recently started working with `livecheck` blocks, successfully create a `strategy` block PR using `#regex`. If someone with limited knowledge of livecheck can do it without any prior help from maintainers, that seems encouraging.

> Also, you have to write `regex` three times instead of simply inlining the regex straight away. Having `regex` outside is also generally one line longer.

Most people are just going to copy-paste one of the examples or an existing `livecheck` block anyway. Anyone else who's bothered by the extra typing can set up snippets (I have a number of livecheck-related ones to save me the trouble of typing the same things repeatedly). These don't seem like compelling reasons to me, especially considering that `strategy` blocks aren't all that common to begin with (~12% of homebrew/cask `livecheck` blocks, less than 1% of homebrew/core).

> Lastly, it will lead to inconsistent usage anyways given there are blocks that need more than one regex.

`strategy` blocks with two regexes only make up around 5% of the existing `strategy` block examples in homebrew/cask (0% in homebrew/core). 89% only use one regex and can be modified to use `#regex` instead. The other 6% don't use a regex at all.

The `strategy` blocks with two regexes are exceptions rather than the norm. It's fine if those don't use `#regex` but the others should. Using `#regex` for `strategy` blocks with one regex is more in line with the 1500+ other `livecheck` blocks that use `#regex` and, at least from my perspective, there's value in surfacing the regex in the debug and JSON output.

**Review by MikeMcQuaid — APPROVED** [review]

**Comment by MikeMcQuaid** [comment]
> Everyone who's working with `livecheck` blocks should already understand the fundamental concept that the regex supplied in the `livecheck` block is provided to the strategy.

I don't think this is a safe assumption.

File: docs/Brew-Livecheck.md
Comment (reitermarkus) [reply docs/Brew-Livecheck.md:70]: Running `brew style --fix` is a required step for both `homebrew/core` and `homebrew/cask`. If you are using an editor which doesn't automatically indent this, you will also likely introduce some sort of other whitespace error, so `brew style --fix` is necessary in most cases anyways.

**Comment by reitermarkus** [comment]
> 89% only use one regex and can be modified to use #regex instead.

That leaves 11% which will be inconsistent with the pattern you are proposing while there will be no ambiguity when putting the regex inside the block 100% of the time.

**Comment by samford** [comment]
> I don't think this is a safe assumption.

If a contributor/maintainer doesn't understand this basic concept about how `livecheck` works with the regex, then they also won't understand how a `strategy` block works. I've been working on `livecheck` for over a year now and I haven't encountered any contributors/maintainers who are confused by how `#regex` functions or have had trouble working with it and most of this was before we had any documentation in place.

I feel that the minor concerns around using `#regex` with a `strategy` block won't come to fruition in practice, since it's defined immediately before the `strategy` block. If there's any potential it's not clear that the regex is the same both outside and inside the `strategy` block, the documentation should clearly explain how this works.

> That leaves 11% which will be inconsistent with the pattern you are proposing while there will be no ambiguity when putting the regex inside the block 100% of the time.

The `strategy` block will also be a black box that we have no insight into from the debug/JSON output 100% of the time instead of only 5% of the time. Using `#regex` allows us to surface that information outside of the 5% that use two regexes. Using a `strategy` block shouldn't require us to lose this information in the output simply because we're unwilling to approach it in a manner that allows the information to continue to be surfaced.

---

As a practical example, if we have an existing `livecheck` block with a `#regex` call and we need to add a `strategy` block to replace certain text in the matched version text, the regex has to be manually moved into the `strategy` block. At this point, people are going to ask, "Why can't I just pass the regex into the `strategy` block? Why do I have to move the regex when all I want to do is replace some text in the version?"

When a `strategy` block only uses one regex, keeping the regex in the block goes against what's normal in `livecheck` blocks that don't use a `strategy` block. This makes it more difficult to modify an existing `livecheck` block to add/remove a `strategy` block. There will also be instances where we need to add a `strategy` block and the generated regex from a strategy is fine but we have to duplicate it within the `strategy` block since we're unwilling to pass in `regex`. As mentioned already, we also lose the ability to surface the regex in the debug/JSON output if it's kept in the `strategy` block.

This all seems like a bad trade-off when all we want is to modify version text. This can be avoided by using `#regex` instead of a regex literal in the `strategy` block. If the only concern is that it may confuse some people, then we can address this with documentation.

**Comment by MikeMcQuaid** [comment]
> > I don't think this is a safe assumption.
> 
> If a contributor/maintainer doesn't understand this basic concept about how `livecheck` works with the regex, then they also won't understand how a `strategy` block works. The documentation should clearly explain how this all works and it should be modified/expanded if it's insufficient.

I don't fully understand this and I've reviewed almost every PR that's come into Homebrew/brew and a few in Homebrew/homebrew-core related to livecheck.

> If the only concern is that it may confuse some people, then we can address this with documentation.

The documentation is already lengthy. I don't expect it's realistic that people are going to be able to jump into the right location to find the answer if they are confused about this.

**Comment by samford** [comment]
> The documentation is already lengthy. I don't expect it's realistic that people are going to be able to jump into the right location to find the answer if they are confused about this.

If they're already looking at the `strategy` block examples in the documentation, then the information would be at the top of that section. I'm talking about maybe one or two more sentences there.

I don't imagine this is going to confuse anyone but I don't think that navigating this documentation will be difficult for anyone who knows enough to contribute to Homebrew repositories.

---

What I'm concerned about is that if we don't use `#regex` (in the majority of cases where it's possible), then `strategy` blocks will be inconsistent with how strategies work. If we're concerned about confusing people, it strikes me that this is much more complicated for people to understand. That is to say, strategies would do one thing but `strategy` blocks would do another.

Contributors/maintainers would need to remember that the regex is passed to strategies but not `strategy` blocks, which is asking more of them.  As mentioned, contributors would also have to manually convert between blocks like:

```ruby
livecheck do
  url :stable
  regex(/href=.*?example[._-]v?(\d+(?:\.\d+)+)\.t/i)
end
```

and

```ruby
livecheck do
  url :stable
  strategy :page_match do |page|
    page.scan(/href=.*?example[._-]v?(\d+(?:\.\d+)+)\.t/i)
        .map { |match| match&.first&.gsub(/\D/, "") }
  end
end
```

If we use `#regex` with the `strategy` block, then we simply add a `strategy` block without having to touch `#regex`:

```ruby
livecheck do
  url :stable
  regex(/href=.*?example[._-]v?(\d+(?:\.\d+)+)\.t/i)
  strategy :page_match do |page|
    page.scan(regex).map { |match| match&.first&.gsub(/\D/, "") }
  end
end
```

There's concern about making this hard for contributors/maintainers but not passing the regex into `strategy` blocks would actively make this more difficult to understand and work with.

**Comment by reitermarkus** [comment]
> the generated regex from a strategy is fine but people are going to naively duplicate the regex in the strategy block (instead of simply passing it in) 

I'd see this as a positive: Using an explicit regex makes the block more readable; when you only see a `regex` variable you cannot reason about what is going on unless you actually run the code.

**Comment by samford** [comment]
> Using an explicit regex makes the block more readable; when you only see a `regex` variable you cannot reason about what is going on unless you actually run the code.

By that logic, we shouldn't have any strategies that generate a URL and regex internally from the provided URL, as that could confuse users unless they run the code. This includes `Apache`, `Bitbucket`, `Cpan`, `GithubLatest`, `Gnome`, `Gnu`, `Hackage`, `Launchpad`, `Npm`, `Pypi`, `Sourceforge`, and `Xorg`.

In these cases, you either have to run the check to understand what happens or you have to look at the code. Clearly we don't want to create a `strategy` block for each of the formulae/casks that use these strategies so why would we want to duplicate the regex that's generated from the strategy when we don't need to?

Passing in the regex is tantamount to saying, "This already works and all I want to do is modify the version text". By passing in the regex, you can modify the version text without needing to understand how the strategy works internally. This is easier for contributors than asking them to understand how the strategy works and to duplicate the regex it uses for a given formula/cask by default.

**Comment by MikeMcQuaid** [comment]
@samford Can we try to find a middle-ground to get this merged?

**Comment by MikeMcQuaid** [comment]
@samford again: can we try to find a way to get this merged? Otherwise I'm tempted to just say this should be merged as-is and any/all improvements should be in a follow-up PR.
@reitermarkus are there any changes you'd want to see before this is merged or are you happy with it as-is?

**Comment by reitermarkus** [comment]
I'm happy with it as-is.

**Comment by MikeMcQuaid** [comment]
Merging as-is for now, thanks @reitermarkus. Let's address any review in a follow-up PR. It may be easier for you to create this @samford when you've got the other open PRs merged out. Thanks all ❤️

**Pull request merged**
